src/librustc_infer/infer/error_reporting/nice_region_error/static_impl_trait.rs

```diff
@@ -2,229 +2,300 @@
-                                        GenericBound::Outlives(Lifetime { name, span, .. })
-                                            if name.ident().to_string() == lifetime_name =>
-                                        {
-                                            Some(*span)
-                                        }
-                                        _ => None,
-                                    })
-                                    .next()
-                                {
-                                } else {
-                                    err.span_suggestion_verbose(
-                                        fn_return.span.shrink_to_hi(),
-                                        &format!(
-                                            "{declare} `impl Trait` {captures}, {explicit}",
-                                            declare = declare,
-                                            captures = captures,
-                                            explicit = explicit,
-                                        ),
-                                        plus_lt.clone(),
-                                        Applicability::MaybeIncorrect,
-                                    );
-                                }
+        // Find all the `impl`s in the local scope that can be called on the type parameter.
+        // FIXME: this doesn't find `impl dyn Trait { /**/ }`.
+        let impl_self_tys = tcx
+            .all_traits(LOCAL_CRATE)
+            .iter()
+            .flat_map(|trait_did| tcx.hir().trait_impls(*trait_did))
+            .filter_map(|impl_node| {
+                let impl_did = tcx.hir().local_def_id(*impl_node);
+                if let Some(Node::Item(Item { kind: ItemKind::Impl { self_ty, .. }, .. })) =
+                    tcx.hir().get_if_local(impl_did.to_def_id())
+                {
+                    Some(self_ty)
+                } else {
+                    None
+                }
+            });
+        let mut suggested = false;
+        for self_ty in impl_self_tys {
+            if let TyKind::TraitObject(
+                poly_trait_refs,
+                Lifetime { name: LifetimeName::ImplicitObjectLifetimeDefault, .. },
+            ) = self_ty.kind
+            {
+                for p in poly_trait_refs {
+                    if let PolyTraitRef {
+                        trait_ref:
+                            TraitRef { path: Path { res: Res::Def(DefKind::Trait, did), .. }, .. },
+                        ..
+                    } = p
+                    {
+                        for found_did in &v.0 {
+                            if did == found_did {
+                                // We've found an `impl Foo for dyn Bar {}`.
+                                // FIXME: we should change this so it also works for
+                                // `impl Foo for Box<dyn Bar> {}`.
+                                err.span_suggestion_verbose(
```

Comment:
This looks like it will lint for every single `impl SomeTrait for dyn TargetTrait` in the current crate, regardless of whether or not they are related to the problematic method call.

I think it would be better to attach additional information in `RegionVariableOrigin` (`AutoRef` in this case) to allow the exact method call to be determined.

Reply:
You are correct. I would like us to do that too, and superficially looked into it, but it seemed like a much bigger change. I'll look into it this weekend.

Reply:
Good news, modifying `AutoRef` to hold `ty::AssocItem` can work. Bad news, in these cases we're dealing with a `MiscVariable`.

Reply:
I managed to carry enough information to only point at the correct `impl`, but not enough to avoid exploring all the available `impl`s for the right one. I also managed to get the item for the `impl dyn Trait {}` case. I am happy with the current impl, if not with the code.